File: src/shapes/circle.js
// Circle shape geometry calculations
export const circleGeometry = {
  // Generate SVG path for circle
  generatePath: (radius) => {
    // Create a circle using line segments (simpler than arc commands)
    const segments = 32; // Number of line segments to approximate circle
    const points = [];
    
    for (let i = 0; i <= segments; i++) {
      const angle = (i / segments) * 2 * Math.PI;
      const x = radius + radius * Math.cos(angle);
      const y = radius + radius * Math.sin(angle);
      points.push(`${i === 0 ? 'M' : 'L'} ${x} ${y}`);
    }
    
    return points.join(' ') + ' Z';
  },

  // Alternative: Use circle element for better performance
  generateCircleElement: (radius) => {
    return {
      type: 'circle',
      cx: radius,
      cy: radius,
      r: radius
    };
  },

  // Calculate bounding box
  getBoundingBox: (radius) => {
    return {
      minX: 0,
      minY: 0,
      maxX: radius * 2,
      maxY: radius * 2,
      width: radius * 2,
      height: radius * 2
    };
  },

  // Calculate area
  getArea: (radius) => {
    return Math.PI * radius * radius;
  },

  // Calculate circumference
  getPerimeter: (radius) => {
    return 2 * Math.PI * radius;
  },

  // Get center point
  getCenter: (radius) => {
    return {
      x: radius,
      y: radius
    };
  },

  // Get vertices for dimension labels
  getVertices: (radius) => {
    // Return key points around the circle for reference
    const points = [];
    const numPoints = 8; // 8 points around the circle
    
    for (let i = 0; i < numPoints; i++) {
      const angle = (i / numPoints) * 2 * Math.PI;
      points.push({
        x: radius + radius * Math.cos(angle),
        y: radius + radius * Math.sin(angle)
      });
    }
    
    return points;
  },
  getDimensionPoints: (radius) => {
    const diameter = radius * 2;
    return {
      diameter: {
        start: { x: 0, y: diameter + 20 },
        end: { x: diameter, y: diameter + 20 },
        label: { x: radius, y: diameter + 35 }
      },
      radius: {
        start: { x: radius, y: radius },
        end: { x: diameter, y: radius },
        label: { x: radius + radius/2, y: radius - 10 }
      },
      radiusVertical: {
        start: { x: radius, y: radius },
        end: { x: radius, y: 0 },
        label: { x: radius + 10, y: radius/2 }
      }
    };
  },

  // Get points at specific angles for reference
  getPointsAtAngles: (radius, angles = [0, 90, 180, 270]) => {
    return angles.map(angle => {
      const radians = (angle * Math.PI) / 180;
      return {
        x: radius + radius * Math.cos(radians),
        y: radius + radius * Math.sin(radians),
        angle
      };
    });
  },

  // Validate parameters
  validateParameters: (radius) => {
    const errors = [];
    
    if (radius <= 0) errors.push('Radius must be greater than 0');
    if (radius > 5000) errors.push('Radius must be less than 5000mm');
    
    return {
      isValid: errors.length === 0,
      errors
    };
  },

  // Get arc path for partial circles (if needed)
  generateArcPath: (radius, startAngle, endAngle) => {
    const startRadians = (startAngle * Math.PI) / 180;
    const endRadians = (endAngle * Math.PI) / 180;
    
    const x1 = radius + radius * Math.cos(startRadians);
    const y1 = radius + radius * Math.sin(startRadians);
    const x2 = radius + radius * Math.cos(endRadians);
    const y2 = radius + radius * Math.sin(endRadians);
    
    const largeArc = endAngle - startAngle > 180 ? 1 : 0;
    
    return `M ${x1} ${y1} A ${radius} ${radius} 0 ${largeArc} 1 ${x2} ${y2}`;
  }
};

export default circleGeometry;

File: src-tauri/src/shapes/circle.rs
use crate::shapes::{BoundingBox, Point, ShapeGeometry, ShapeParameters, ValidationError, Dimension, DimensionOrientation, Transform};



pub struct CircleGeometry;



impl ShapeGeometry for CircleGeometry {

    fn generate_path(&self, params: &ShapeParameters) -> String {

        let radius = params.radius.unwrap_or(50.0);

        let segments = 32;

        let mut points = Vec::new();

        

        for i in 0..=segments {

            let angle = (i as f64 / segments as f64) * 2.0 * std::f64::consts::PI;

            let x = radius + radius * angle.cos();

            let y = radius + radius * angle.sin();

            

            if i == 0 {

                points.push(format!("M {} {}", x, y));

            } else {

                points.push(format!("L {} {}", x, y));

            }

        }

        

        format!("{} Z", points.join(" "))

    }



    fn get_bounding_box(&self, params: &ShapeParameters) -> BoundingBox {

        let radius = params.radius.unwrap_or(50.0);

        let diameter = radius * 2.0;

        

        BoundingBox {

            min_x: 0.0,

            min_y: 0.0,

            max_x: diameter,

            max_y: diameter,

            width: diameter,

            height: diameter,

        }

    }



    fn get_area(&self, params: &ShapeParameters) -> f64 {

        let radius = params.radius.unwrap_or(50.0);

        std::f64::consts::PI * radius * radius

    }



    fn get_perimeter(&self, params: &ShapeParameters) -> f64 {

        let radius = params.radius.unwrap_or(50.0);

        2.0 * std::f64::consts::PI * radius

    }



    fn get_center(&self, params: &ShapeParameters) -> Point {

        let radius = params.radius.unwrap_or(50.0);

        Point {

            x: radius,

            y: radius,

        }

    }



    fn get_vertices(&self, params: &ShapeParameters) -> Vec<Point> {

        let radius = params.radius.unwrap_or(50.0);

        let num_points = 8;

        let mut points = Vec::new();

        

        for i in 0..num_points {

            let angle = (i as f64 / num_points as f64) * 2.0 * std::f64::consts::PI;

            points.push(Point {

                x: radius + radius * angle.cos(),

                y: radius + radius * angle.sin(),

            });

        }

        

        points

    }



    fn validate_parameters(&self, params: &ShapeParameters) -> ValidationError {

        let mut errors = Vec::new();

        

        if let Some(radius) = params.radius {

            if radius <= 0.0 {

                errors.push("Radius must be greater than 0".to_string());

            }

            if radius > 5000.0 {

                errors.push("Radius must be less than 5000mm".to_string());

            }

        } else {

            errors.push("Radius is required".to_string());

        }

        

        ValidationError {

            is_valid: errors.is_empty(),

            errors,

        }

    }



    fn get_dimensions(&self, params: &ShapeParameters, render_offset: &Point, transform: &Transform) -> Vec<Dimension> {

        let radius = params.radius.unwrap_or(50.0);

        let diameter = radius * 2.0;

        let center = Point { x: radius + render_offset.x, y: radius + render_offset.y };

        

        // Dynamic offset calculation - 8% of dimension with bounds

        let diameter_offset = (diameter * 0.08).max(5.0).min(30.0);

        

        vec![

            // Diameter dimension (vertical through center)

            Dimension {

                start_point: Point { x: radius + render_offset.x, y: diameter_offset + render_offset.y },

                end_point: Point { x: radius + render_offset.x, y: diameter - diameter_offset + render_offset.y },

                text_position: Point { x: radius + diameter_offset + (diameter_offset * 0.75) + render_offset.x, y: radius + render_offset.y },

                value: diameter,

                label: format!("Ø {:.0}mm", diameter),

                orientation: DimensionOrientation::Vertical,

            },

        ]

    }

    fn transform_point(&self, point: &Point, center: &Point, transform: &Transform) -> Point {
        let mut x = point.x;
        let mut y = point.y;

        // Apply rotation
        if transform.rotation != 0.0 {
            let angle_rad = (transform.rotation * std::f64::consts::PI) / 180.0;
            let cos_a = angle_rad.cos();
            let sin_a = angle_rad.sin();
            
            let rel_x = x - center.x;
            let rel_y = y - center.y;
            
            x = rel_x * cos_a - rel_y * sin_a + center.x;
            y = rel_x * sin_a + rel_y * cos_a + center.y;
        }

        // Apply flips
        if transform.flip_x {
            x = 2.0 * center.x - x;
        }
        if transform.flip_y {
            y = 2.0 * center.y - y;
        }

        Point { x, y }
    }

    fn get_rotation_center(&self, params: &ShapeParameters) -> Point {
        let radius = params.radius.unwrap_or(50.0);
        Point { x: radius, y: radius }
    }

}

